fabtests/common/shared.c

```diff
@@ -401,6 +401,15 @@ void ft_fill_mr_attr(struct iovec *iov, int iov_count, uint64_t access,
 	}
 }
 
+bool ft_need_mr_reg(struct fi_info *fi)
+{
+	return (((fi->domain_attr->mr_mode & FI_MR_LOCAL) ||
+		  (opts.options & FT_OPT_USE_DEVICE)) &&
+		 ((fi->domain_attr->mr_mode & FI_MR_HMEM) ||
+		  !(opts.options & FT_OPT_USE_DEVICE))) ||
+		(fi->caps & (FI_RMA | FI_ATOMIC));
```

Comment:
I know this is the same logic as before (just negated), but is this correct?

The last part -- using RMA -- makes sense.  The other checks look backwards.

For example, combining FI_MR_LOCAL with using device memory _looks_ wrong. FI_MR_LOCAL is independent of device buffers.  The next part of the check is also confusing.  FI_MR_HMEM being OR'ed against NOT using device buffers is unexpected.  FI_MR_HMEM only applies to device buffers.  The following check seem more natural to me and are easier to understand:

(fi->domain_attr->mr_mode & FI_MR_LOCAL) ||
(fi->caps & (FI_RMA | FI_ATOMIC) ||
((fi->domain_attr->mr_mode & FI_MR_HMEM) &&
 (opts.options & FT_OPT_USE_DEVICE))

Sooo... *if* the above checks *don't* work, why not?

Reply:
I *think* the intended logic of the old code was something like this: we need memory registration when (1) RMA or Atomic capabilities are present (2) FT_OPT_USE_DEVICE and FI_MR_HMEM are set (3) FT_OPT_USE_DEVICE is not set and FI_MR_LOCAL is set. Since most of our providers support RMA and/or Atomic, we probably never ran the other paths.

I've updated it with the logic you propose.